src/test/webServer.ts

```diff
@@ -0,0 +1,140 @@
+};
+
+const newProcessLogPrefixer = () => new stream.Transform({
+  transform(this: stream.Transform, chunk: any, encoding: string, callback: stream.TransformCallback) {
+    this.push(chunk.toString().split(os.EOL).map((line: string): string => line ? `[Playwright WebServer] ${line}` : line).join(os.EOL));
+    callback();
+  },
+});
+
+export class WebServer {
+  private _process?: ChildProcess;
+  private _port?: number;
+  private _processExitedWithNonZeroStatusCode!: Promise<any>;
+  private _config: Omit<WebServerConfig, 'port'>;
+  constructor(config: WebServerConfig) {
+    this._config = config;
+    this._port = config.port;
+  }
+
+  public static async create(config: WebServerConfig): Promise<WebServer> {
+    const webServer = new WebServer(config);
+    if (webServer._port)
+      await webServer._verifyFreePort();
+    try {
+      await webServer._startWebServer();
+      await webServer._waitForAvailability();
+    } catch (error) {
+      await webServer.kill();
+      throw error;
+    }
+    process.env.PLAYWRIGHT_WEB_SERVER_URL = `http://localhost:${webServer._port}`;
+    return webServer;
+  }
+
+  private async _verifyFreePort() {
+    assert(this._port);
+    const cancellationToken = { canceled: false };
+    const portIsUsed = await Promise.race([
+      new Promise(resolve => setTimeout(() => resolve(false), 100)),
+      waitForSocket(this._port, 100, cancellationToken),
+    ]);
+    cancellationToken.canceled = true;
+    if (portIsUsed)
+      throw new Error(`Port ${this._port} is used, make sure that nothing is running on the port`);
+  }
+
+  private async _startWebServer(): Promise<void> {
+    let collectPortResolve = (port: number) => { };
+    const collectPortPromise = new Promise<number>(resolve => collectPortResolve = resolve);
+    function collectPort(data: Buffer) {
+      const regExp = /http:\/\/localhost:(\d+)/.exec(data.toString());
+      if (regExp)
+        collectPortResolve(parseInt(regExp[1], 10));
+    }
+
+    this._process = exec(this._config.command, {
+      env: {
+        ...DEFAULT_ENVIRONMENT_VARIABLES,
+        ...process.env,
+        ...this._config.env,
+      },
+    });
+    this._process.stdout.pipe(newProcessLogPrefixer()).pipe(process.stdout);
+    this._process.stderr.pipe(newProcessLogPrefixer()).pipe(process.stderr);
+    if (!this._port)
+      this._process.stdout.on('data', collectPort);
+    let processExitedWithNonZeroStatusCodeCallback = (error: Error) => { };
+    this._processExitedWithNonZeroStatusCode = new Promise((_, reject) => processExitedWithNonZeroStatusCodeCallback = reject);
+    this._process.on('exit', code => processExitedWithNonZeroStatusCodeCallback(new Error(`WebServer was not able to start. Exit code: ${code}`)));
+    if (!this._port) {
+      this._port = await collectPortPromise;
+      console.log(`[Playwright Test] Port ${this._port} by process '${this._config.command}' was automatically detected. Starting the tests...`);
```

Comment:
Why the prefix? We never prefix anything else in our messages.

I'd also drop "starting the tests..." because you never know what's next in the pipeline :) Instead, you can print "waiting for the port to respond" if it takes more than 500(?) ms.

Reply:
Since most of the users are using web frameworks it mostly takes long, thats why I'm printing now the "detecting the port" message always.